viewing_party/party.py

```diff
@@ -1,23 +1,128 @@
+        return 0.0
+    
+    for i in range(len(user_data["watched"])):
+            total_ratings += user_data["watched"][i]["rating"]
+            average_rating = total_ratings / number_movies
+    return average_rating
+
+
+def get_most_watched_genre(user_data):
+    genre_options = []
+
+    if user_data["watched"] == []:
+        return None    
+    
+    for i in range(len(user_data["watched"])):
+        genre_options.append(user_data["watched"][i]["genre"])
+        highest_watched = (max(set(genre_options), key=genre_options.count))
+    return highest_watched
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data): 
+    unique_watch = []
+    friends_movies_watched = []
+
+    for i in range(len(user_data["friends"])):
+        for movie in user_data["friends"][i]["watched"]:
+            friends_movies_watched.append(movie)
+ 
+    for movie in user_data["watched"]:
+        if movie not in friends_movies_watched:
+            unique_watch.append(movie)
+    return unique_watch
+
+
+def get_friends_unique_watched(user_data):
+    unique_watch = []
+    no_duplicate_unique_watch = []
+   
+
+    for i in range(len(user_data["friends"])):
+        for movie in user_data["friends"][i]["watched"]:
+            if movie not in user_data["watched"]:
+                    unique_watch.append(movie)
+
+    for movie in unique_watch:
+        if movie not in no_duplicate_unique_watch:
+            no_duplicate_unique_watch.append(movie)
+    return no_duplicate_unique_watch
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    available_recs = []
+    friends_movies_recs = get_friends_unique_watched(user_data)
+
+    for i in range(len(friends_movies_recs)):
+        if friends_movies_recs[i]["host"] in user_data["subscriptions"]:
+            available_recs.append(friends_movies_recs[i])
+    return available_recs
+        
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):    
+    available_recs = []
+    friends_movies_recs = get_friends_unique_watched(user_data)
+    user_fav_genre = get_most_watched_genre(user_data)
+
+    for i in range(len(friends_movies_recs)):
+        if friends_movies_recs[i]["genre"] == user_fav_genre:
+            available_recs.append(friends_movies_recs[i])
+    return available_recs
+
+
+def get_rec_from_favorites(user_data):
+    available_recs = []
+    users_watchlist = get_unique_watched(user_data)
+
+    for i in range(len(user_data["favorites"])):
+        if user_data["favorites"][i] in users_watchlist:
+            available_recs.append(user_data["favorites"][i])
+    return available_recs

```

Comment:
Same thing I've been saying before for each of these last two functions as well about the for loops, but other than that, you used helper functions really well!